src/devices/cpu/e132xs/e132xsop.hxx

```diff
@@ -1311,191 +1250,248 @@ void hyperstone_device::hyperstone_add_global_global()
-	const int64_t tmp = (int64_t)sreg + (int64_t)dreg;
-
-	SR &= ~(V_MASK | Z_MASK | N_MASK);
-	SR |= ((sreg ^ tmp) & (dreg ^ tmp) & 0x80000000) >> 28;
-
-//#if SETCARRYS
-//  CHECK_C(tmp);
-//#endif
+	check_delay_PC();
 
-	const int32_t res = sreg + dreg;
-	m_global_regs[dst_code] = res;
+	const uint32_t dreg = ~(SRC_GLOBAL ? m_global_regs[SRC_CODE] : m_local_regs[(SRC_CODE + GET_FP) & 0x3f]);
+	if (DST_GLOBAL)
+		set_global_register(DST_CODE, dreg);
+	else
+		m_local_regs[(DST_CODE + GET_FP) & 0x3f] = dreg;
 
-	if (res == 0)
+	if (dreg == 0)
 		SR |= Z_MASK;
-	SR |= SIGN_TO_N(res);
+	else
+		SR &= ~Z_MASK;
 
 	m_icount -= m_clock_cycles_1;
-
-	if (SR & V_MASK)
-		execute_exception(get_trap_addr(TRAPNO_RANGE_ERROR));
 }
 
-void hyperstone_device::hyperstone_adds_local_global()
+template <hyperstone_device::reg_bank DST_GLOBAL, hyperstone_device::reg_bank SRC_GLOBAL>
+void hyperstone_device::hyperstone_sub()
 {
-	const uint32_t src_code = SRC_CODE;
-	const uint32_t dst_code = (DST_CODE + GET_FP) & 0x3f;
-	const int32_t sreg = (src_code == SR_REGISTER) ? (SR & C_MASK) : (int32_t)m_global_regs[src_code];
-	const int32_t dreg = m_local_regs[dst_code];
-	const int64_t tmp = (int64_t)sreg + (int64_t)dreg;
+	check_delay_PC();
 
-	SR &= ~(V_MASK | Z_MASK | N_MASK);
-	SR |= ((sreg ^ tmp) & (dreg ^ tmp) & 0x80000000) >> 28;
+	const uint32_t fp = GET_FP;
+	const uint32_t src_code = SRC_GLOBAL ? SRC_CODE : ((SRC_CODE + fp) & 0x3f);
+	const uint32_t dst_code = DST_GLOBAL ? DST_CODE : ((DST_CODE + fp) & 0x3f);
 
-//#if SETCARRYS
-//  CHECK_C(tmp);
-//#endif
+	const uint32_t sreg = SRC_GLOBAL ? ((src_code == SR_REGISTER) ? GET_C : m_global_regs[src_code]) : m_local_regs[src_code];
+	uint32_t dreg = (DST_GLOBAL ? m_global_regs : m_local_regs)[dst_code];
 
-	const int32_t res = sreg + dreg;
-	m_local_regs[dst_code] = res;
+	const uint64_t tmp = uint64_t(dreg) - uint64_t(sreg);
 
-	if (res == 0)
+	SR &= ~(C_MASK | V_MASK | Z_MASK | N_MASK);
+
+	SR |= (tmp & 0x100000000L) >> 32;
+	SR |= ((tmp ^ dreg) & (dreg ^ sreg) & 0x80000000) >> 28;
+
+	dreg -= sreg;
+	if (DST_GLOBAL)
+	{
+		set_global_register(dst_code, dreg);
+
+		if (dst_code == PC_REGISTER)
+			SR &= ~M_MASK;
+	}
+	else
+	{
+		m_local_regs[dst_code] = dreg;
+	}
+
+	if (dreg == 0)
 		SR |= Z_MASK;
-	SR |= SIGN_TO_N(res);
+	SR |= SIGN_TO_N(dreg);
 
 	m_icount -= m_clock_cycles_1;
-
-	if (SR & V_MASK)
-		execute_exception(get_trap_addr(TRAPNO_RANGE_ERROR));
 }
 
-void hyperstone_device::hyperstone_adds_local_local()
+template <hyperstone_device::reg_bank DST_GLOBAL, hyperstone_device::reg_bank SRC_GLOBAL>
+void hyperstone_device::hyperstone_subs()
 {
 	const uint32_t fp = GET_FP;
-	const uint32_t dst_code = (DST_CODE + fp) & 0x3f;
-	const int32_t sreg = (int32_t)m_local_regs[(SRC_CODE + fp) & 0x3f];
-	const int32_t dreg = m_local_regs[dst_code];
-	const int64_t tmp = (int64_t)sreg + (int64_t)dreg;
+	const uint32_t src_code = SRC_GLOBAL ? SRC_CODE : ((SRC_CODE + fp) & 0x3f);
```

Comment:
There's no call to `check_delay_PC` here - the similar negs handler does have such a call. Is this intentional or an oversight?